### ISSUE TITLE
mirrored text on lcd

### ISSUE THREAD

**Issue by aadhyatmik** [fruit-bat/pico-zxspectrum#115]
Great project, kudos to you, when I tried to assemble on a zero pcb to test, I get mirrored text on lcd, using zxspectrumpicomputerzxlcd uf2 file gp2-gp6 for ili9341 lcd and gp19-gp13 for sdmicro. Someone please suggest a fix.
![IMG_20240203_230648](https://github.com/fruit-bat/pico-zxspectrum/assets/158766924/8d576863-18f9-4027-8abb-09fa892a0f15)

**Comment by fruit-bat** [comment]
That will be because the driver is for a st7789. 

A tweak to the MADCTL (Memory Access Control 0x36) register will probably sort the rotation. Possibly 0x47 or 0x87.

Fixing the colours may be more involved... but interesting to try!

[ILI9341_DS_V1.10_20110415.pdf](https://github.com/fruit-bat/pico-zxspectrum/files/14152593/ILI9341_DS_V1.10_20110415.pdf)

See: https://github.com/fruit-bat/pico-zxspectrum/blob/main/src/picomputer/picomputer_st7789/st7789_lcd.cpp
 
You in a position to build the firmware?

**Comment by fruit-bat** [comment]
Looks like blue and red are swapped.  That what you think?

**Comment by aadhyatmik** [comment]
Thanks for your reply, I have never built, however i have picosdk installed, will surely try, kindly guide me stepwise.

I think I need to replace tinyusb then tweak madctl value ( kindly suggest filename), then execute cmake command in build folder then execute make command.

**Comment by fruit-bat** [comment]
Yup, I can have a go for you if you like... give me a few minutes....

**Comment by aadhyatmik** [comment]
Unfortunately I am not good at recognising colours, it seems a washed out to me. Will see once it's working.

**Comment by fruit-bat** [comment]
This is 0x10, so we know for the next try.

[ZxSpectrumPicomputerZxLcd.uf2.zip](https://github.com/fruit-bat/pico-zxspectrum/files/14152725/ZxSpectrumPicomputerZxLcd.uf2.zip)

NB you need to unzip the uf2

**Comment by aadhyatmik** [comment]
Unfortunately that makes it behave more erratic.

![IMG_20240204_013415](https://github.com/fruit-bat/pico-zxspectrum/assets/158766924/56061d34-aa36-4d30-9903-dd523928516a)

**Comment by aadhyatmik** [comment]
X30 seems good, attached picture for reference, colours also look right to me ;-)

![IMG_20240204_014518](https://github.com/fruit-bat/pico-zxspectrum/assets/158766924/0abf24e8-f8c7-4183-8b55-d72531d02275)

**Comment by aadhyatmik** [comment]
Thank you very much, I need some more time to test usb keyboard etc. I am preparing this for my daughter, she love programming. 🙏

**Comment by fruit-bat** [comment]
Colours look inverted (or wrong in some other way_ to me...

What does this one look like?

[ZxSpectrumPicomputerZxLcd.uf2.zip](https://github.com/fruit-bat/pico-zxspectrum/files/14154200/ZxSpectrumPicomputerZxLcd.uf2.zip)

**Comment by aadhyatmik** [comment]
X30 looked better to me, stripes on logo in front of this are blue cyan green yellow, previous one (X30) contains yellow red pink blue.

![IMG_20240204_020505](https://github.com/fruit-bat/pico-zxspectrum/assets/158766924/62023b01-5770-48cb-8627-cad9ee6060d3)

**Comment by aadhyatmik** [comment]
X30 looked better to me, stripes on logo in front of this are blue cyan green yellow, previous one (X30) contains yellow red pink blue.

![IMG_20240204_020505](https://github.com/fruit-bat/pico-zxspectrum/assets/158766924/62023b01-5770-48cb-8627-cad9ee6060d3)

**Comment by aadhyatmik** [comment]
Kindly tell me when you are done, will close it.

**Comment by fruit-bat** [comment]
This swaps the red and blue components...

[ZxSpectrumPicomputerZxLcd.uf2.zip](https://github.com/fruit-bat/pico-zxspectrum/files/14154360/ZxSpectrumPicomputerZxLcd.uf2.zip)

Hopefully,  that should do it :-)

**Comment by aadhyatmik** [comment]
Sorry for delay, this one I think matches with the logo, colours are red, yellow, green and cyan.

![Uploading IMG_20240204_064159.jpg…]()

**Comment by aadhyatmik** [comment]
Here is the picture
![IMG_20240204_064159](https://github.com/fruit-bat/pico-zxspectrum/assets/158766924/2a2e57a9-bdd6-4131-bc27-23b12ae9c482)

**Comment by aadhyatmik** [comment]
On testing keyboard it is working but I was not able to type " or : and few other keys , tried 2 different keyboards, one dell wired keyboard and one portronics hydra in wired configuration, it seems a layout issue may be. Seems need to read manual.

**Comment by fruit-bat** [comment]
The keys on the USB keyboard map 'literally' to the spectrum keys. Alt-gr (on the right) is symbol shift.

**Comment by aadhyatmik** [comment]
Thanks will try alt-gr tonight, never owned a spectrum or any 8 bit computer, so missed that. 🙏
